intrusive_collections_examples/src/basic_datastructures/red_black_tree.rs

```diff
@@ -0,0 +1,359 @@
+/// }
+///
+/// fn main() {
+///     let thing = Thing::default();
+///     let ptr1: *const Thing = &thing;
+///
+///     // We could use any field we wanted, we're using `parent` just as an
+///     // example here.
+///     let ptr2: *const RedBlackTreeNode = &(thing.parent);
+///     let offset = unsafe { (ptr2 as *const u8).offset_from(ptr1 as *const u8) };
+///     println!(
+///         "The 'parent' field is offset from the start of 'thing' by {:?} bytes",
+///         offset
+///     );
+///     assert!(offset >= 0);
+///     let ptr3: *const Thing = unsafe { ((ptr2 as *const u8).sub(offset as usize)) as *const Thing };
+///     assert_eq!(ptr1, ptr3);
+/// }
+/// ```
+///
+/// As you can see, we're doing a fair amount of casting, and there is some
+/// `unsafe` code within the example.  This is a cost of intrusive collections,
+/// and one that you should be aware of when implementing them.  Which brings up
+/// the talk about safety...
+///
+/// ## Safety, or, how to shoot yourself in the foot
+///
+/// Intrusive collections are generally unsafe to work with.  They require
+/// working with raw pointers instead of references, and this can't easily be
+/// changed.  For a more in-depth explanation of why this is so, read the
+/// [`README.md`](../README.md) file.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct RedBlackTreeNode<T>
+where
+    T: Debug + Clone + Default + PartialEq + Eq + PartialOrd + Ord + Hash,
+{
+    /// This is the key used to compare different nodes in the tree.  While it
+    /// is possible to defer this to the object that owns the node, we're going
+    /// to use the simpler route and store the key in here.
+    key: T,
+
+    /// All red-black tree nodes are colored either red or black.
+    ///
+    /// Our chosen default color is black, which means that when we create a
+    /// single root node, it is a complete tree by itself.
+    color: RedBlackTreeNodeColor,
+
+    /// The node that is considered to be the parent of this node.  If there is
+    /// no parent, then this will be `None`.
+    parent: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the left child of this node.  If there
+    /// is no left child, then this will be `None`.
+    left_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// The node that is considered to be the right child of this node.  If
+    /// there is no right child, then this will be `None`.
+    right_child: Option<*const RedBlackTreeNode<T>>,
+
+    /// Since we're using raw pointers, we **cannot** allow these instances to
+    /// be moved by the compiler (dropping is OK, but moving will almost
+    /// certainly cause UB).  For more information on why this is necessary,
+    /// start reading at
+    /// https://doc.rust-lang.org/std/pin/index.html#example-self-referential-struct
+    /// and be extra careful to read all the way through the explanations about
+    /// drop!
+    _pin: PhantomPinned,
```

Comment:
Since intrusive collections can't provide a safe interface, this doesn't matter as much here. But Putting `PhantomPinned` is a nice default.

Reply:
I've never been clear on rust's move semantics.  I keep worrying that the compiler will be able to copy an instance from one location to another, invalidating some raw pointers.  Since dereferencing a raw pointer is unsafe, the compiler throws the onus of preventing this on the programmer.  I'm just trying to prevent potential errors... but I honestly don't know if it's necessary here!

Reply:
Well, `Pin` doesn't actually have any special meaning to the compiler, so it's not going to change anything the compiler does 😄. It's a pure library construct that uses clever `unsafe` to prevent access to a `&mut T` (where `T: !Unpin`) with safe code. Nothing special. Rust will never move anything behind your back, but it is rather easy to move things so your concerns are well-founded.

I would stick the containing node into a box, then the only way to move it would be to dereference the box. (But this has other concerns, we may need an `AliasableBox` because currently `Box` behaves very similarly to `&mut T`, which we can't allow for intrusive collections).

Reply:
> Well, Pin doesn't actually have any special meaning to the compiler, so it's not going to change anything the compiler does :smile:. It's a pure library construct that uses clever `unsafe` to prevent access to a `&mut T` (where `T: !Unpin`) with safe code. Nothing special. Rust will never move anything behind your back, but it is rather easy to move things so your concerns are well-founded.

Good to know, but I keep having these nightmarish visions of the stack getting reordered behind my back, or closures doing something weird... I know I shouldn't, but `move` *just as a word* terrifies me! :fearful:!  Honestly, I know that it's 100% impossible in rust, I just can't shake the feeling that my raw pointers are going to stop pointing at what I think they're point to because something got moved that shouldn't...

> I would stick the containing node into a box, then the only way to move it would be to dereference the box. (But this has other concerns, we may need an `AliasableBox` because currently Box behaves very similarly to `&mut T`, which we can't allow for intrusive collections).

You're right, but I'm going to defer working on that until after I have a complete red-black tree done, with tests in place.